core/win32/callback.c

```diff
@@ -5647,6 +5647,46 @@ intercept_exception(app_state_at_intercept_t *state)
              */
             mcontext_to_context(cxt, &mcontext, false /* !set_cur_seg */);
 
+            /*
+             * i#2144 : we check if this is a single step exception
+             * where we diverted the address.
+             */
+            if (pExcptRec->ExceptionCode == EXCEPTION_SINGLE_STEP) {
+                instr_t instr;
+
+                instr_init(dcontext, &instr);
+                decode(dcontext, pExcptRec->ExceptionAddress, &instr);
+                /* Checks that exception address translate on a popf. */
+                if (instr_get_opcode(&instr) == OP_popf) {
+                    instr_reset(dcontext, &instr);
+                    decode(dcontext, faulting_pc, &instr);
+                    /* Checks that exceptionwas generated by a nop. */
+                    if (instr_get_opcode(&instr) == OP_nop) {
+                        LOG(THREAD, LOG_ASYNCH, 2,
+                            "Caught generated single step exception at "PFX"\n",
+                            pExcptRec->ExceptionAddress);
+                        /* Will continue after one byte popf. */
+                        dcontext->next_tag = mcontext.pc + 1;
```

Comment:
Is this the right way to of before calling `transfer_to_dispatch` ? (setting next_tag)